habitat_baselines/common/obs_transformers.py

```diff
@@ -217,166 +218,676 @@ def from_config(cls, config: Config):
+
+    def _generate_grid_one_output(
+        self, output_model: CameraProjection
+    ) -> torch.Tensor:
+        # Obtain points on unit sphere
+        world_pts, not_assigned_mask = output_model.unprojection()
         # Generate grid
         grids = []
-        for rot in rotations:
-            R = torch.from_numpy(rot.T).float()
-            rotate_on_sphere = torch.matmul(
-                xyz_on_sphere.view((-1, 3)), R
-            ).view(equ_h, equ_w, 3)
-
-            # Project points on z=1 plane
-            grid = rotate_on_sphere / torch.abs(rotate_on_sphere[..., 2:3])
-            mask = torch.abs(grid).max(-1)[0] <= 1  # -1 <= grid.xy <= 1
-            mask *= grid[..., 2] == 1
-            grid[
-                ~mask
-            ] = 2  # values bigger than one will be ignored by grid_sample
-            grid_xy = -grid[..., :2].unsqueeze(0)
-            grids.append(grid_xy)
-        grids = torch.cat(grids, dim=0)
+        for input_model in self.input_models:
+            grid, input_mask = input_model.projection(world_pts)
+            # Make sure each point is only assigned to single input
+            input_mask *= not_assigned_mask
+            # Values bigger than one will be ignored by grid_sample
+            grid[~input_mask] = 2
+            # Update not_assigned_mask
+            not_assigned_mask *= ~input_mask
+            grids.append(grid)
+        grids = torch.stack(grids, dim=0)
         return grids
 
-    def _to_equirec(self, batch: torch.Tensor) -> torch.Tensor:
-        """Takes a batch of cubemaps stacked in proper order and converts thems to equirects, reduces batch size by 6"""
+    def generate_grid(self) -> torch.Tensor:
+        multi_output_grids = []
+        for output_model in self.output_models:
+            grids = self._generate_grid_one_output(output_model)
+            multi_output_grids.append(grids)
+        multi_output_grids = torch.stack(multi_output_grids, dim=0)
+        return multi_output_grids  # output_len, input_len, output_img_h, output_img_w, 2
+
+    def _convert(self, batch: torch.Tensor) -> torch.Tensor:
+        """Takes a batch of images stacked in proper order and converts thems, reduces batch size by input_len"""
         batch_size, ch, _H, _W = batch.shape
-        if batch_size == 0 or batch_size % 6 != 0:
-            raise ValueError("Batch size should be 6x")
+        out_h, out_w = self.output_models[0].size()
+        if batch_size == 0 or batch_size % self.input_len != 0:
+            raise ValueError(f"Batch size should be {self.input_len}x")
         output = torch.nn.functional.grid_sample(
             batch,
             self._grids_cache,
             align_corners=True,
             padding_mode="zeros",
         )
         output = output.view(
-            batch_size // 6, 6, ch, self.equ_h, self.equ_w
-        ).sum(dim=1)
-        return output  # batch_size // 6, ch, self.equ_h, self.equ_w
+            self.input_len,
```

Comment:
Hold on, does this still work? Was this tested on out output with VectorEnv? Want to make sure swapping the view ordering doesn't break anything. In this case of batch size 1, this should work (1/6 vs 6/1, but want to make sure it still works on higher batch sizes.)

Reply:
I didn't aware of `VectorEnv`. Let me check if it works.

Reply:
I see, I totally misunderstood the order, If there are two batches from two envs `(R_1st, G_1st, B_1st)` and `(R_2nd, G_2nd, B_2nd)`, the order should be `[R_1st, R_2nd, G_1st, G_2nd, B_1st, B_2nd]` not that `[R_1st, G_1st, B_1st, R_2nd, G_2nd, B_2nd]`. I will fix it.

Reply:
Oh I see, I need to fix the order for output,,, Currently, the order is `[R_1st, G_1st, B_1st, R_2nd, G_2nd, B_2nd]`.

Reply:
I fixed it. I think it works now.